### PR TITLE
✨ Generic enemy

### LINKED ISSUE #268: 🤖 Make Enemies more generic

### PR DESCRIPTION
**Pull request by 429277**
## Content
Reworked enemy code structure to make it more generic.

## Changes
`Assets/Scripts/Units/Axe Unit/AxeUnit.cs` moved code to base
`Assets/Scripts/Units/Axe Unit/AxeUnitAttackState.cs` updated methods
`Assets/Scripts/Units/Axe Unit/AxeUnitChaseState.cs` updated methods
`Assets/Scripts/Units/Axe Unit/AxeUnitDeathState.cs` moved code to generic behaviour
`Assets/Scripts/Units/Axe Unit/AxeUnitIdleState.cs` updated methods
`Assets/Scripts/Units/Axe Unit/AxeUnitStunnedState.cs` moved code to generic behaviour
`Assets/Scripts/Units/Axe Unit/IAxeUnitState.cs` moved code to generic behaviour
`Assets/Scripts/Units/Javelin Unit/IJavelinUnitState.cs` moved code to generic behaviour
`Assets/Scripts/Units/Javelin Unit/JavelinUnit.cs` moved code to generic behaviour
`Assets/Scripts/Units/Javelin Unit/JavelinUnitAttackState.cs` updated methods
`Assets/Scripts/Units/Javelin Unit/JavelinUnitChaseState.cs` updated methods
`Assets/Scripts/Units/Javelin Unit/JavelinUnitDeathState.cs` moved code to generic behaviour
`Assets/Scripts/Units/Javelin Unit/JavelinUnitIdleState.cs` updated methods
`Assets/Scripts/Units/Javelin Unit/JavelinUnitStunnedState.cs` moved code to generic behaviour

### Added
`Assets/Scripts/Units/DeathBehaviour.cs` created generic behaviour
`Assets/Scripts/Units/Enemy.cs` added generic data
`Assets/Scripts/Units/GenericFiniteStateMachine.cs` created generic behaviour
`Assets/Scripts/Units/IState.cs` created generic behaviour
`Assets/Scripts/Units/StunnedBehaviour.cs` created generic behaviour

## Testing
I tested if enemies still work in game. Both enemies still work as intended. 

Closes #268

### CODE REVIEW

File: Assets/Scripts/Units/StunnedBehaviour.cs [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+        private float _timer;
+        public Enemy Unit { set; private get; }
+
+        public void StartStun()
+        {
+            _timer = 0;
+            Unit.LocalAnimator.SetTrigger("Stunned");
+        }
+
+        public bool StunEnded()
```
Comment (NathanThus) [inline Assets/Scripts/Units/StunnedBehaviour.cs:18]: Any particular reason you're not using `UniTask.WaitForSeconds(time)`

File: Assets/Scripts/Units/Enemy.cs [hunk 2]
```diff
@@ -7,7 +7,46 @@ namespace RogueApeStudio.Crusader.Units
...
+        public Animator LocalAnimator => _localAnimator;
+        public float StartAttackDistance => _startAttackDistance;
+        public float DestroyTime => _destroyTime;
+        public float SelfStunDuration => _selfStunDuration;
+        public float DetectionRange => _detectionRange;
+
         public abstract bool Stun();
         public abstract bool IsDead();
+
+        public void DestroySelf()
```
Comment (NathanThus) [inline Assets/Scripts/Units/Enemy.cs:36]: Isn't this a job for the `EnemyHealth` script? Why does the pathfinding/combat system need to destroy itself?

Comment (429277) [reply Assets/Scripts/Units/Enemy.cs:36]: Its an enemy behaviour script, the health script does not contain any functionality for death.

File: Assets/Scripts/Units/StunnedBehaviour.cs (hunk 1 above)
Comment (429277) [reply Assets/Scripts/Units/StunnedBehaviour.cs:18]: No reason, this just works.

File: Assets/Scripts/Units/Enemy.cs (hunk 2 above)
Comment (NathanThus) [reply Assets/Scripts/Units/Enemy.cs:36]: Sure, I can see that.
I'd have preferred to see it in `EnemyHealth`, but sure.

File: Assets/Scripts/Units/StunnedBehaviour.cs (hunk 1 above)
Comment (NathanThus) [reply Assets/Scripts/Units/StunnedBehaviour.cs:18]: I'd prefer it if you used `UniTask` in the future, but I'll let it through for now.

**Review by NathanThus — APPROVED** [review]

**Review by davey2206 — APPROVED** [review]

**Pull request merged**